les/utils/limiter.go

```diff
@@ -0,0 +1,405 @@
+// selection round before removed, with a weight based on the last relative cost.
+// If no more requests are added to the queue before it is selected again then the
+// queue is removed and the next time a request comes from the same node the queue is
+// added with the highest possible weight.
+func (l *Limiter) choose() *nodeQueue {
+	if l.valueSelect.IsEmpty() || l.lastValueSelected {
+		if ag, ok := l.addressSelect.Choose().(*addressGroup); ok {
+			l.lastValueSelected = false
+			return ag.choose()
+		}
+	}
+	nq, _ := l.valueSelect.Choose().(*nodeQueue)
+	l.lastValueSelected = true
+	return nq
+}
+
+// processLoop processes requests sequentially
+func (l *Limiter) processLoop() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for {
+		if l.quit {
+			for _, nq := range l.nodes {
+				for _, request := range nq.queue {
+					close(request.process)
+				}
+			}
+			return
+		}
+		nq := l.choose()
+		if nq == nil {
+			l.cond.Wait()
+			continue
+		}
+		if len(nq.queue) > 0 {
+			request := nq.queue[0]
+			nq.queue = nq.queue[1:]
+			nq.totalPriorWeight -= request.priorWeight
+			l.totalPriorWeight -= request.priorWeight
+			nq.relCost = 0
+			l.lock.Unlock()
+			costCh := make(chan float64)
+			request.process <- costCh
+			relCost := <-costCh / float64(l.maxPriorWeight)
+			l.lock.Lock()
+			nq.relCost += relCost
+			l.update(nq)
+		} else {
+			l.remove(nq)
+		}
+	}
+}
+
+// Stop stops the processing loop. All queued and future requests are rejected.
+func (l *Limiter) Stop() {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	l.quit = true
+	l.cond.Signal()
+}
+
+type (
+	dropList     []dropListItem
+	dropListItem struct {
+		nq       *nodeQueue
+		priority float64
+	}
+)
+
+func (l dropList) Len() int {
+	return len(l)
+}
+
+func (l dropList) Less(i, j int) bool {
+	return l[i].priority < l[j].priority
+}
+
+func (l dropList) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+
+// dropRequests selects the nodes with the highest queued prior weight to selection
+// weight ratio and drops their queued request. The empty node queues stay in the
+// selectors with a low selection weight in order to penalize these nodes.
+func (l *Limiter) dropRequests() {
+	var (
+		sumValue float64
+		list     dropList
+	)
+	for _, nq := range l.nodes {
+		sumValue += nq.value
+	}
+	for _, nq := range l.nodes {
+		if nq.totalPriorWeight == 0 {
+			continue
+		}
+		w := 1 / float64(len(l.addresses)*len(l.addresses[nq.address].nodes))
+		if sumValue > 0 {
+			w += nq.value / sumValue
+		}
+		list = append(list, dropListItem{
+			nq:       nq,
+			priority: w / float64(nq.totalPriorWeight),
+		})
+	}
+	sort.Sort(list)
+	for _, item := range list {
+		for _, request := range item.nq.queue {
+			close(request.process)
+		}
+		l.totalPriorWeight -= item.nq.totalPriorWeight
+		// set the last relative cost as if all dropped requests were processed with
+		// the highest possible cost (which equals their prior weight).
+		item.nq.relCost += float64(item.nq.totalPriorWeight)
```

Comment:
Why to add the `totalPriorWeight` directly? In another place https://github.com/ethereum/go-ethereum/pull/21676/files#diff-05f453344a1370c706aca53fb8864362a541daeca84a5c82aecf9344c7af20acR326 it's divided by the base

Reply:
Right, it was a mistake, now it is simplified and handled correctly.